clang/lib/Lex/LiteralSupport.cpp

```diff
@@ -1127,9 +1148,9 @@ NumericLiteralParser::NumericLiteralParser(StringRef TokSpelling,
       // wb and WB are allowed, but a mixture of cases like Wb or wB is not. We
       // explicitly do not support the suffix in C++ as an extension because a
       // library-based UDL that resolves to a library type may be more
-      // appropriate there.
-      if (!LangOpts.CPlusPlus && ((s[0] == 'w' && s[1] == 'b') ||
-          (s[0] == 'W' && s[1] == 'B'))) {
+      // appropriate there. The same rules apply for __wb/__WB.
+      if ((!LangOpts.CPlusPlus || PossibleBitInt) &&
+          ((s[0] == 'w' && s[1] == 'b') || (s[0] == 'W' && s[1] == 'B'))) {
```

Comment:
I think there's an existing bug here when `s + 1 >= ThisTokEnd` that should be guarded for (and a test case added).

Reply:
Would the existing test cases of `1__w` and `1.0_` in `bitint-constants.cpp` be enough to test this and the similar above comment?

Reply:
In `Sema::ActOnNumericConstant()`:
>   // NumericLiteralParser wants to overread by one character.  Add padding to
  // the buffer in case the token is copied to the buffer.  If getSpelling()
  // returns a StringRef to the memory buffer, it should have a null char at
  // the EOF, so it is also safe.

Reply:
So it seems that the intention here really is that `NumericLiteralParser` should always be passed a null-terminated string.

Reply:
I think we should look at `ThisTokEnd`. Assuming there is something past it is, at best, looking for trouble.
(If we don't want to do a comprehensive fix here, we should add an assert)

Reply:
Sorry a little confused here, are we keeping the `s + 1 < ThisTokEnd` checks or should we change them to `s < ThisTokEnd` to allow looking into `ThisTokEnd` (for example`s < ThisTokEnd && s[1] == '_'`)